server/src/main/java/org/candlepin/async/JobManager.java

```diff
@@ -495,6 +500,30 @@ public synchronized void initialize() throws StateManagementException {
         }
     }
 
+    /**
+     * Attempts to recover all jobs that were abandoned by this node if it shutdown abnormally
+     * while executing jobs. Any jobs in the RUNNING state with an executor set to this node
+     * will have their state forcefully rewound to QUEUED in an attempt to allow the task to be
+     * rerun.
+     */
+    private void recoverAbandonedJobs() {
+        AsyncJobStatusQueryBuilder queryBuilder = new AsyncJobStatusQueryBuilder()
+            .setJobStates(Collections.singleton(JobState.RUNNING))
+            .setExecutors(Collections.singleton(Util.getHostname()));
+
+        // Impl note: this violates normal state transitions, but we're trying to recover from an
+        // ungraceful shutdown in which our process was terminated while a job was running. We'll
+        // set the job's state to QUEUED, which will allow us to pick up the job and run it again
+        // next time the message is received. If the message has been lost, then the job cleaner
+        // will eventually nuke this job as part of its non-terminal job aborting step.
+        for (AsyncJobStatus job : this.jobCurator.findJobs(queryBuilder)) {
+            log.warn("Recovering abandoned job: {}", job);
+
+            job.setState(JobState.QUEUED);
+            this.jobCurator.merge(job);
```

Comment:
Same question here: why fetch job objects into memory? we don't even do transition validation here, so an update query would be much faster

Reply:
The number here *should* be no more than the number of job executor threads configured (default is 10), so it shouldn't be a *huge* hit compared to a raw query. The important thing I want here, though, is the logging.

Reply:
You're right. 10 jobs should be negligible